### DISCUSSION TITLE
"setBackgroundImage()" & "enableEffect()" functions is not present in "@antmedia/webrtc_adaptor" library in angular( version 13)

### DISCUSSION THREAD

**Discussion by swatisinha-mis** [ant-media/Ant-Media-Server#5498] [Q&A]
I am facing problem for publishing the stream with virtual background Image and also with enabling effect on video. But When I am trying to run "https://github.com/ant-media/StreamApp/blob/master/src/main/webapp/publish_webrtc_virtual_background.html" code sample in my project, I received "Cannot read properties of undefined (reading 'setBackgroundImage')" & "Cannot read properties of undefined (reading 'enableEffect')" error. Please help me to resolve this problem.

Thankyou

**Comment by USAMAWIZARD** [comment]
Can you provide some code snippets how are you trying to call those functions.

**Reply by swatisinha-mis** [reply]
// HTML Code

<div class="row mt-4">
    <div class="col-md-12" style="text-align: center;">
      <div id="my_camera" style="width:100%;">
        <p>
          <video id="localVideo" autoplay muted controls playsinline width="560" height="368"></video>
          <img id="virtualBackgroundImage" src="../../../assets/images/test/virtual-background.png" hidden />

        </p>
      </div>
      <button (click)="startStreaming()" class="btn btn-success me-2">Start</button>
      <button (click)="StopPublishing()" class="btn btn-danger">End</button>
    </div>

    <div class="col-sm-12 text-center" id="virtualBackgrounds">
      <h3>Select a virtual background Image</h3>
      <div class="d-flex" id="virtualbackgroundimages">
        <img src="../../../assets/images/test/noeffect-background.png" id="noeffect"
          class="backgroundImages selected" />
        <img src="../../../assets/images/test/blur-background.png" id="blur" class="backgroundImages " />
        <img src="../../../assets/images/test/virtual-background.png" id="antMediaBackground"
          class="backgroundImages " />
        <img src="../../../assets/images/test/cloud-background.png" class="backgroundImages" />

        <img src="../../../assets/images/test/bgt1.jpg" class="backgroundImages" />
        <img src="../../../assets/images/test/bgt2.jpg" class="backgroundImages" />
        <img src="../../../assets/images/test/bgt3.jpg" class="backgroundImages" />
        <img src="../../../assets/images/test/bgt4.png" class="backgroundImages" />
      </div>

    </div>
  </div>



//Typescript code

import { WebRTCAdaptor } from '@antmedia/webrtc_adaptor';
import { VideoEffect } from '@antmedia/webrtc_adaptor';
declare var $: any;

  title = 'antmedialivetest';

  /* antmedia variable */
  //webrtc
  webrtcAdaptor!: any;
  token: any;
  pc_config: any = null;
  sdpConstraintsPublish: any;
  mediaConstraintsPublish: any;
  websocketURL!: string;
  streamId: any;
  infodata: string = '';
  messagedata: any;
  isSend: boolean = false;

  canvas: any;
  video: any;
  localCameraView: any;

  constructor() {
    this.streamId = "aHtrfu9eFupp1674472507702";
    this.webrtcAdaptor = null;

    // ant media     
    this.token = "";

    this.sdpConstraintsPublish = {
      OfferToReceiveAudio: false,
      OfferToReceiveVideo: false
    };

    this.mediaConstraintsPublish = {
      video: true,
      audio: true
    };
  }

  goLive() {
    this.isSend = true;
    this.streamId = "aHtrfu9eFupp1674472507702";
    this.PublishStreaming(this.streamId);
  }

  initWebrtc() {
    this.webrtcAdaptor = new WebRTCAdaptor({
      websocket_url: this.websocketURL,
      mediaConstraints: this.mediaConstraintsPublish,
      peerconnection_config: this.pc_config,
      sdp_constraints: this.sdpConstraintsPublish,
      localVideoId: "localVideo",
      // localStream: localStream,
      // remoteVideoId: "localVideo",
      bandwidth: 900,
      debug: true,
      isPlayMode: true,
      dataChannelEnabled: true,
      callback: (info: any, description: any) => {
        this.infodata = info;
        if (info == "initialized") {
        } else if (info == "pong") {
        } else if (info == "publish_started") {
          console.log("publish started");
        } else if (info == "publish_finished") {
          console.log("publish finished");
        } else if (info == "data_received") {
          console.log(description);
        } else if (info == "available_devices") {
        } else if (info == "closed") {
          if (typeof description != "undefined") {
            this.webrtcAdaptor = null;
            this.initWebrtc();
          }
        }
      },
      callbackError: (error: any, message: any) => {
        console.log("error callback: " + JSON.stringify(error));        
      }
    });
  }
  
  ngAfterViewInit(): void {
    this.localCameraView = null;
    navigator.mediaDevices.getUserMedia({ audio: true, video: true }).then((stream: any) => {
      this.localCameraView = document.getElementById("localVideo");
      this.localCameraView.srcObject = stream;
      this.localCameraView.play();
      this.video = document.querySelector('video#localVideo');
      this.initWebrtc();
    });
  }

  PublishStreaming(streamId: any) {
    this.webrtcAdaptor.publish(streamId, "", "", "", "", "", "");
  }

  startStreaming() {
    this.webrtcAdaptor.publish(this.streamId, "", "", "", "", "", "");
  }

  StopPublishing() {
    this.isSend = false;
    this.webrtcAdaptor.stop(this.streamId);
  }

  ngOnInit(): void {
    $(".backgroundImages").click(this.enableVirtualBackground);
  }

  enableVirtualBackground() {
    $(".backgroundImages").removeClass("selected");
    var effectName;
    if ($(this).attr("id") == "blur") {
      effectName = VideoEffect.BLUR_BACKGROUND;
    }
    else if ($(this).attr("id") == "noeffect") {
      effectName = VideoEffect.NO_EFFECT;
    }
    else {
      effectName = VideoEffect.VIRTUAL_BACKGROUND;
      this.webrtcAdaptor.setBackgroundImage(this);
    }

    this.webrtcAdaptor.enableEffect(effectName).then(() => {
      // console.log("Effect: "+ effectName+" is enabled");
    }).catch((err: any) => {
      // console.error("Effect: "+ effectName+" is not enabled. Error is " + err);
      console.error(err);
    });
    $(this).addClass("selected");
  }

**Reply by USAMAWIZARD** [reply]
i don't have much idea about angular is it possible for you to share a small  (minimal )project in which this problem is reproducible.

**Reply by rahul78275** [reply]
Hello @swatisinha-mis,  I can try in weekend if i get the component files (.ts, style and html). Also share the package.json file.

**Reply by rahul78275** [reply]
![image](https://github.com/ant-media/Ant-Media-Server/assets/36676540/72fd0f71-4b60-4586-8e23-405d4d34be6d)

**Reply by swatisinha-mis** [reply]
is it in angular? Can you share your code to my mail id "swatisinhamis@gmail.com".

**Comment by swatisinha-mis** [comment]
> i don't have much idea about angular is it possible for you to share a small (minimal )project in which this problem is reproducible.

Hello, 
 "https://drive.google.com/file/d/1kFYv0CxK4KMO-A63SUC6xhqdiuPpavLv/view?usp=sharing" is my sample project link, please check this one.

**Comment by rahul7827** [comment]
> is it in angular? Can you share your code to my mail id "[swatisinhamis@gmail.com](mailto:swatisinhamis@gmail.com)".

Yes, It is angular.

// Problem is here
`$(".backgroundImages").click(this.enableVirtualBackground); `


// In template you can do like this
`<img src="../../../assets/images/test/noeffect-background.png" id="noeffect" class="backgroundImages selected" (click)="enableVirtualBackground('noeffect', $event)" />`


//update your ngAfterViewInit hook like this, no need to use navigator.mediaDevices.getUserMedia({ audio: true, video: true }), 
```
ngAfterViewInit(): void {
    this.initWebrtc();
}
```

**Reply by USAMAWIZARD** [reply]
Thanks @rahul7827  for help

**Comment by swatisinha-mis** [comment]
> > is it in angular? Can you share your code to my mail id "[swatisinhamis@gmail.com](mailto:swatisinhamis@gmail.com)".
> 
> Yes, It is angular.
> 
> // Problem is here `$(".backgroundImages").click(this.enableVirtualBackground); `
> 
> // In template you can do like this `<img src="../../../assets/images/test/noeffect-background.png" id="noeffect" class="backgroundImages selected" (click)="enableVirtualBackground('noeffect', $event)" />`
> 
> //update your ngAfterViewInit hook like this, no need to use navigator.mediaDevices.getUserMedia({ audio: true, video: true }),
> 
> ```
> ngAfterViewInit(): void {
>     this.initWebrtc();
> }
> ```

I still facing this problem see the attached screenshot below.
![2023-09-06](https://github.com/ant-media/Ant-Media-Server/assets/99073345/2bbe8344-a4f8-45c8-acb1-6e21c8307483)
can you please share your component files.

**Reply by rahul78275** [reply]
Hello @swatisinha-mis 

I have already sent you the component files yesterday, Please confirm your email (swatisinhamis@gmail.com). Check it into spam also.

![image](https://github.com/ant-media/Ant-Media-Server/assets/36676540/3e907f82-8ca0-4b9e-8447-0489606c243c)

**Reply by USAMAWIZARD** [reply]
@swatisinha-mis  which version of webrtc-adapter library are you using ?

**Reply by rahul78275** [reply]
[app.component.zip](https://github.com/ant-media/Ant-Media-Server/files/12537245/app.component.zip)

**Reply by swatisinha-mis** [reply]
Thankyou so much for the code, but it's not working with "@antmedia/webrtc_adaptor": "^2.6.3-SNAPSHOT-2023-Jun-26-08-23" version of webrtc-adapter library

**Reply by rahul78275** [reply]
Ashambhav! It is working fine at my end, you can check my stream at your media server end (easestream). You can schedule a google meet call if you wish.

**Reply by swatisinha-mis** [reply]
Yes, you can send me google meet link in my gmail account.

**Reply by rahul78275** [reply]
Sent.

**Reply by yogeshJain0727** [reply]
@rahul78275 could you please send the entire project so that we can check what may be causing issues with our code?

**Reply by USAMAWIZARD** [reply]
@yogeshJain0727  can you please check if you are using same webrtcadapter as rahul  
Can you please try with this.
npm i @antmedia/webrtc_adaptor@2.7.0-SNAPSHOT-2023-Aug-27-08-14

**Reply by rahul7827** [reply]
@USAMAWIZARD  I am using the same webrtc adaptor as swati sent to me.

**Reply by rahul78275** [reply]
Hi @yogeshJain0727 

I have shared the entire project on mail. You can check it. BTW, I guess issue is not related to source code. This may be dependency issue.
Also do share the error message which we get during npm start, this will help us to understand the issue in detail.

**Comment by rahul7827** [comment]
> @rahul78275 could you please send the entire project so that we can check what may be causing issues with our code?

I had used the same code which swati sent to me (even node_modules files are same). I also tried to regenerate the issue by creating fresh **npm i**. I am away from my system. Will send you the complete project today night.

Can you please try checking this project on another machine.

**Reply by yogeshJain0727** [reply]
ok
